pkg/api/volume/formatter.go

```diff
@@ -0,0 +1,83 @@
+package volume
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	v1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+)
+
+const (
+	actionExport = "export"
+)
+
+func Formatter(request *types.APIRequest, resource *types.RawResource) {
+	resource.Actions = make(map[string]string, 1)
+	if resource.APIObject.Data().String("spec", "sourceType") == harvesterv1.VirtualMachineImageSourceTypeExportVolume {
+		resource.AddAction(request, actionExport)
+	}
+}
+
+type ExportActionHandler struct {
+	images   v1beta1.VirtualMachineImageClient
+	pvcCache v1.PersistentVolumeClaimCache
+}
+
+func (h ExportActionHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if err := h.do(rw, req); err != nil {
+		status := http.StatusInternalServerError
+		if e, ok := err.(*apierror.APIError); ok {
+			status = e.Code.Status
+		}
+		rw.WriteHeader(status)
+		_, _ = rw.Write([]byte(err.Error()))
+		return
+	}
+	rw.WriteHeader(http.StatusOK)
+}
+
+func (h ExportActionHandler) do(rw http.ResponseWriter, r *http.Request) error {
+	vars := mux.Vars(r)
+	action := vars["action"]
+	namespace := vars["namespace"]
+	displayName := vars["name"]
+	pvcName := vars["pvc"]
+
+	switch action {
+	case actionExport:
+		return h.exportVolume(r.Context(), namespace, displayName, pvcName)
+	default:
+		return apierror.NewAPIError(validation.InvalidAction, "Unsupported action")
+	}
+}
+
+func (h ExportActionHandler) exportVolume(ctx context.Context, namespace, displayName, pvcName string) error {
+	// We only permit volume source from existing PersistentVolumeClaim at this moment.
+	// KubeVirt won't check PVC existence so we validate it on our own.
+	pvc, err := h.pvcCache.Get(namespace, pvcName)
+	if err != nil {
+		return fmt.Errorf("failed to get pvc %s/%s, error: %s", pvcName, namespace, err.Error())
+	}
```

Comment:
Move this to the validation webhook(pkg/webhook/resources/virtualmachineimage/validator.go)

Reply:
@gitlawr fixed

Reply:
@gitlawr updated